### PR TITLE
First pass at XWayland <-> Wayland copy/paste

### PR DESCRIPTION
**Pull request by wmww**
Still lots of rough edges, but this does work in the simple cases.

Limitations:
- Pasting blocks the Wayland thread
- No support for incremental operations

I intend to address these before I consider the feature complete, but I don't see why any of them would stop this being reviewed and merged. I think it's better to merge stuff as I go rather than have one monster PR at the end.

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
Tried this. With kate running on X11 and kate running on Wayland.

Copy-paste of a small block of text worked from Wayland to X11
Copy-paste of a small block of text FAILED from X11 to Wayland with
```plain
Data set on unsupported clipboard mode. QMimeData object will be deleted.
```

**Comment by wmww** [comment]
Hm, thanks. I was able to reproduce.

That warning doesn't appear when run on Sway, but does always appear on Mir (even when XWayland is disabled). It also seems to appear on select (not copy or paste), which indicates to me it stems from us not supporting primary selection, and is thus unrelated.

I don't think this issue needs to block this PR (lots of other stuff doesn't work as well), but I might fix it today. If you need something to test with that does work I've been using gedit as well as miscellaneous other things like Blender and PCManFM.

**Comment by wmww** [comment]
Actually, the results I'm getting are more complicated:
- Wayland Kate -> Wayland Kate: seems to work
- X11 Kate -> X11 Kate: 
- Wayland Kate -> X11 Kate: Works if the copy is made before the the X server has started, doesn't work with copies made after. Except sometimes it does.
- Wayland Gnome Terminal -> X11 Kate: Works, but fails if the previous copy was from Wayland Kate. Copying twice in a row (even without changing the selection or window focus in-between) always works.
- X11 Kate -> Wayland Kate: Sometimes works, sometimes doesn't. When it fails, I seem to be getting `xwayland: can not copy from X11 because operation is currently in progress` at time of attempted paste, which is a significant clue.

**Comment by wmww** [comment]
This is too messy, here's an exact reproducer for one specific bug (this is mainly for my benefit): 
1. Open Wayland Kate
2. Open X11 Kate
3. Copy the same thing from Wayland Kate twice
4. Paste into X11 Kate (or X11 gedit)
5. Doesn't work

Modifications that cause it to work as expected:
- Copying twice from Wayland Kate *before* X11 Kate opens
- Only coping once from Wayland Kate
- Copying twice from Wayland Gnome Terminal
- Pasting into any Wayland app

**Comment by wmww** [comment]
I definitely fixed a problem and whacking at it for a bit I didn't stumble across any more. Let me know if you find anything.

**Review by AlanGriffiths — APPROVED** [review]
A couple of nits, but working well enough to merge.

bors r=wmww

File: src/server/frontend_xwayland/xwayland_clipboard_provider.cpp [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+namespace mf = mir::frontend;
+namespace ms = mir::scene;
+
+namespace
+{
+size_t const INCREMENT_CHUNK_SIZE = 64 * 1024;
```
[the remark this answers is not in the document]
Comment (wmww) [reply src/server/frontend_xwayland/xwayland_clipboard_provider.cpp:35]: Been writing Rust, forgot we don't do constants like this.

**Review by AlanGriffiths — APPROVED** [review]
Good now

bors r+

**Pull request merged**